Public/Src/Pips/Dll/Operations/PipGraphFragmentSerializer.cs

```diff
@@ -0,0 +1,88 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics.ContractsLight;
+using System.IO;
+using BuildXL.Utilities;
+
+namespace BuildXL.Pips.Operations
+{
+    /// <summary>
+    /// Class to handle serialization/deserialization of pip graph fragments
+    /// </summary>
+    public class PipGraphFragmentSerializer
+    {
+        /// <summary>
+        /// Total pips in the fragment.  Until this is read in from the file, it is 0.
+        /// </summary>
+        public int TotalPips { get; private set; }
+
+        /// <summary>
+        /// Total pips deserialized so far.
+        /// </summary>
+        public int PipsDeserialized { get; private set; }
+
+        /// <summary>
+        /// Total pips serialized so far.
+        /// </summary>
+        public int PipsSerialized { get; private set; }
+
+        /// <summary>
+        /// Name of the fragment, for printing on the console
+        /// </summary>
+        public string FragmentName { get; private set; }
+
+        /// <summary>
+        /// Deserialize a pip graph fragment and call the given handleDeserializedPip function on each pip deserialized
+        /// Returns true if successfully handled all pips.
+        /// </summary>
+        public bool Deserialize(string fragmentName, PipExecutionContext context, PipGraphFragmentContext pipFragmentContext, AbsolutePath filePath, Func<Pip, bool> handleDeserializedPip)
+        {
+            FragmentName = fragmentName;
+            PipsDeserialized = 0;
+            string fileName = filePath.ToString(context.PathTable);
+            Contract.Assert(File.Exists(fileName), "Pip graph fragment file doesn't exist");

```

Comment:
I don't think this is a useful assertion.  If the file doesn't exist, `new FileStream()` will immediately blow up and will have the same affect as failing this assertion check.

Reply:
I guess I was just going for a nicer error message, but I can remove it.

---
In reply to: [289556967](https://github.com/microsoft/BuildXL/pull/394#discussion_r289556967) [](ancestors = 289556967)